boot/bootstate16.go

```diff
@@ -0,0 +1,152 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/snap"
+)
+
+type bootState16 struct {
+	varSuffix string
+	errName   string
+}
+
+func newBootState16(typ snap.Type) *bootState16 {
+	var varSuffix, errName string
+	switch typ {
+	case snap.TypeKernel:
+		varSuffix = "kernel"
+		errName = "kernel"
+	case snap.TypeBase:
+		varSuffix = "core"
+		errName = "boot base"
+	default:
+		panic(fmt.Sprintf("cannot make a bootState16 for snap type %q", typ))
+	}
+	return &bootState16{varSuffix: varSuffix, errName: errName}
+}
+
+func (s *bootState16) revisions() (snap, try_snap *NameAndRevision, trying bool, err error) {
+	bloader, err := bootloader.Find("", nil)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot get boot settings: %s", err)
+	}
+
+	snapVar := "snap_" + s.varSuffix
+	trySnapVar := "snap_try_" + s.varSuffix
+	vars := []string{"snap_mode", snapVar, trySnapVar}
+	snaps := make(map[string]*NameAndRevision, 2)
+
+	m, err := bloader.GetBootVars(vars...)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot get boot variables: %s", err)
+	}
+
+	for _, vName := range vars {
+		v := m[vName]
+		if v == "" && vName != snapVar {
+			// snap_mode & snap_try_<type> can be empty
+			// snap_<type> cannot be! and will fail parsing
+			// below
+			continue
+		}
+
+		if vName == "snap_mode" {
+			trying = v == "trying"
+		} else {
+			nameAndRevno, err := nameAndRevnoFromSnap(v)
+			if err != nil {
+				return nil, nil, false, fmt.Errorf("cannot get name and revision of %s (%s): %v", s.errName, vName, err)
+			}
+			snaps[vName] = nameAndRevno
+		}
+	}
+
+	return snaps[snapVar], snaps[trySnapVar], trying, nil
+}
+
+type bootStateUpdate16 struct {
+	bl       bootloader.Bootloader
+	env      map[string]string
+	toCommit map[string]string
+}
+
+func newBootStateUpdate16(u bootStateUpdate, names ...string) (*bootStateUpdate16, error) {
+	if u != nil {
+		u16, ok := u.(*bootStateUpdate16)
+		if !ok {
+			return nil, fmt.Errorf("internal error: threading unexpected boot state update: %T", u)
+		}
+		return u16, nil
+	}
+	bl, err := bootloader.Find("", nil)
+	if err != nil {
+		return nil, err
+	}
+	m, err := bl.GetBootVars(names...)
+	if err != nil {
+		return nil, err
+	}
+	return &bootStateUpdate16{bl: bl, env: m, toCommit: make(map[string]string)}, nil
+}
+
+func (u16 *bootStateUpdate16) commit() error {
+	if len(u16.toCommit) == 0 {
+		// nothing to do
+		return nil
+	}
+	env := u16.env
+	// TODO: we could just SetBootVars(toCommit) but it's not
+	// fully backward compatible with the preexisting behavior
+	for k, v := range u16.toCommit {
+		env[k] = v
+	}
+	return u16.bl.SetBootVars(env)
+}
+
+func (s *bootState16) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
```

Comment:
(nitpick^2) maybe `s *bootState16 -> s16 *bootState16` here to be consistent with the commit() func above?

Reply:
I'll see about this in the last one of the chain